### ISSUE TITLE
 An attacker can potentially extend the hash length without knowing the original data, thereby bypassing the Merkle proof verification hence unauthorized addresses being able to access functionalities reserved for whitelisted lenders

### ISSUE THREAD

**Issue by Odhiambo526** [shieldify-security/ion-protocol-c7bf309e9dba7e19eb81e456fa4cd3254e5e65b9#24]
### Code Snippet:
[Whitelist.sol#L118-L137](https://github.com/shieldify-security/ion-protocol-c7bf309e9dba7e19eb81e456fa4cd3254e5e65b9/blob/e417da59cbc7c4e7b2b7643055d21154b29fe292/src/Whitelist.sol#L118-L137)
### Impact:
The impact of this vulnerability is severe. If successfully exploited, an attacker could bypass the whitelist verification process and gain unauthorized access to lender functionalities. This could lead to unauthorized lending activities or manipulation of the lending system.
### Vulnerability Detail:
The vulnerability lies within the `isWhitelistedLender` function of the Whitelist contract. Let's break down the vulnerability step by step:
- The `isWhitelistedLender` function is designed to verify if an address is whitelisted as a lender. It takes in parameters such as the caller address (`poolCaller`), the address to be checked (`addr`), and a Merkle proof (`proof`).
The vulnerability stems from how the leaf hash is constructed within the function:
```solidity
bytes32 leaf = keccak256(bytes.concat(keccak256(abi.encode(addr))));
```
Here's what's happening:
- The address (`addr`) is encoded using `abi.encode`.
- The resulting bytes are concatenated with an empty byte array using `bytes.concat`.
- The concatenated bytes are hashed using `keccak256` to produce the leaf hash.
The vulnerability exposes the contract to a potential hash extension vulnerability. In this case, an attacker exploits certain hash functions' properties to append additional data to the hash without knowing the original data. This means an attacker could potentially manipulate the hash output without knowing the original address, which could lead to bypassing the Merkle proof verification.
### Example of Attack Scenario:
An attacker may observe the construction of the leaf hash and realize that it only depends on the address (`addr`). With this knowledge, the attacker could potentially craft a modified hash by extending the hash length without knowing the original address. This modified hash could then be used to bypass the whitelist verification and gain unauthorized access.
### Severity
In summary, the vulnerability in the Whitelist contract poses a high risk due to the insecure construction of the leaf hash in the `isWhitelistedLender` function. This vulnerability exposes the contract to potential attacks where an unauthorized address could bypass whitelist verification, potentially leading to unauthorized access to lender functionalities and compromising the integrity of the lending system.

**Comment by 0xShax2nk** [comment]
Whitelist is OOS.

**Comment by junkim012** [comment]
Please provide a PoC on how extending the hash length can enable the attacker to bypass the whitelist. It is unclear from the explanation how this attack would actually be carried out in practice.
